File: examples/go/auction/Auction.tsop.go
package contract

import "tsop"

type Auction struct {
	tsop.StatefulSmartContract
	Auctioneer    tsop.PubKey `tsop:"readonly"`
	HighestBidder tsop.PubKey // stateful
	HighestBid    tsop.Bigint // stateful
	Deadline      tsop.Bigint `tsop:"readonly"` // block height deadline
}

// State-mutating: compiler auto-injects checkPreimage + state continuation
func (c *Auction) Bid(bidder tsop.PubKey, bidAmount tsop.Bigint) {
	// Bid must be higher than current highest
	tsop.Assert(bidAmount > c.HighestBid)

	// Auction must not have ended
	tsop.Assert(tsop.ExtractLocktime(c.TxPreimage) < c.Deadline)

	// Update state
	c.HighestBidder = bidder
	c.HighestBid = bidAmount
}

// Non-mutating: compiler auto-injects checkPreimage only (no state continuation)
func (c *Auction) Close(sig tsop.Sig) {
	// Only auctioneer can close
	tsop.Assert(tsop.CheckSig(sig, c.Auctioneer))

	// Auction must have ended
	tsop.Assert(tsop.ExtractLocktime(c.TxPreimage) >= c.Deadline)
}

File: examples/rust/auction/Auction.tsop.rs
use tsop::prelude::*;

#[tsop::contract]
pub struct Auction {
    #[readonly]
    pub auctioneer: PubKey,
    pub highest_bidder: PubKey, // stateful
    pub highest_bid: Bigint,    // stateful
    #[readonly]
    pub deadline: Bigint, // block height deadline
    pub tx_preimage: SigHashPreimage,
}

#[tsop::methods(Auction)]
impl Auction {
    /// State-mutating: compiler auto-injects checkPreimage + state continuation
    #[public]
    pub fn bid(&mut self, bidder: PubKey, bid_amount: Bigint) {
        assert!(bid_amount > self.highest_bid);
        assert!(extract_locktime(&self.tx_preimage) < self.deadline);
        self.highest_bidder = bidder;
        self.highest_bid = bid_amount;
    }

    /// Non-mutating: compiler auto-injects checkPreimage only
    #[public]
    pub fn close(&self, sig: &Sig) {
        assert!(check_sig(sig, &self.auctioneer));
        assert!(extract_locktime(&self.tx_preimage) >= self.deadline);
    }
}
